backend/app/model/mixins/representative_file_version.rb

```diff
@@ -124,26 +124,38 @@ def thumbnail_use_statement_id
       thumbnail_use_statement_id
     end
 
-    # this is an expensive operation for large trees with no representative.
-    # perhaps there is a way to do this without iterating through query results?
     def find_representative_in_resource_tree(record_set, parent_id = nil)
-      same_parent_set = record_set.filter(archival_object__parent_id: parent_id)
-      return nil if same_parent_set.count == 0
-
-      same_parent_set.each do |row|
-        # does this node have a representative?
+      id_positions = {nil => 0}
+      id_depths = {nil => 0}
+      parent_to_child_id = {}
+      id_digital_objects = {}
+
+      record_set.each do |row|
+        child_id = row[:archival_object_id]
+        position = row[:archival_object_position]
+        parent_id = row[:archival_object_parent_id]
+        id_positions[child_id] ||= position
+        parent_to_child_id[parent_id] ||= []
+        parent_to_child_id[parent_id] << child_id
         if row[:digital_object_id] && row[:digital_object_is_representative]
-          return row[:digital_object_id]
+          id_digital_objects[child_id] = row[:digital_object_id]
         end
+      end
 
-        # does this node have a descendent with a representative?
-        if (result = find_representative_in_resource_tree(record_set, row[:archival_object_id]))
-          return result
+      root_set = [nil]
+
```

Comment:
I had typed up a big thing trying to outline what happens here, but after I did all that I realized the question I should ask is: does `id_digital_objects` or `parent_to_child_id` start out with a value associated to `nil`? If not, it seems like nothing could ever happen in this loop, since `next_rec` start out as `nil`. 😅

Reply:
All the top-level nodes in the tree have `nil` as their parent id, so there will be contents in the array `parent_to_child_id[nil]`, and `id_digital_objects[nil]` could have contents or not depending on the data.

Reply:
AH ok, got it.